internal/neofs/epoch.go

```diff
@@ -2,72 +2,168 @@ package neofs
 	}
 
-	// PeriodicGetter implements [EpochGetter].
-	PeriodicGetter struct {
-		logger    *zap.Logger
-		netGetter NetworkInfoGetter
-		epoch     atomic.Uint64
-		timeOut   time.Duration
+	// EpochListener implements [EpochGetter].
+	EpochListener struct {
+		log            *zap.Logger
+		epoch          atomic.Uint64
+		client         *rpcclient.WSClient
+		notifyChan     chan *state.ContainedNotificationEvent
+		netMapContract util.Uint160
+		endpoints      []string
 	}
 )
 
-// NewPeriodicGetter is a constructor to [PeriodicGetter].
-func NewPeriodicGetter(ctx context.Context, initialEpoch uint64, timeOut time.Duration, netGetter NetworkInfoGetter, logger *zap.Logger) *PeriodicGetter {
-	getter := &PeriodicGetter{
-		timeOut:   timeOut,
-		netGetter: netGetter,
-		logger:    logger,
-	}
-
-	getter.epoch.Store(initialEpoch)
-
-	go getter.update(ctx)
+var (
+	eventName = "NewEpoch"
+)
 
-	return getter
+// NewEpochListener is a constructor for EpochListener.
+func NewEpochListener(endpoints []string, log *zap.Logger, netMapContract util.Uint160) *EpochListener {
+	return &EpochListener{
+		log:            log.With(zap.String("service", "epoch_notifications")),
+		endpoints:      endpoints,
+		netMapContract: netMapContract,
+	}
 }
 
 // CurrentEpoch returns actual epoch.
 //
 // CurrentEpoch implements [EpochGetter].
-func (g *PeriodicGetter) CurrentEpoch() uint64 {
-	return g.epoch.Load()
+func (l *EpochListener) CurrentEpoch() uint64 {
+	return l.epoch.Load()
+}
+
+func (l *EpochListener) forceGetEpoch() error {
+	inv := invoker.New(l.client, nil)
+
+	netMapReader := netmap.NewReader(inv, l.netMapContract)
+	epoch, err := netMapReader.Epoch()
+	if err != nil {
+		return fmt.Errorf("failed to get actual epoch: %w", err)
+	}
+
+	l.updateEpoch(epoch.Uint64())
+	return nil
 }
 
-func (g *PeriodicGetter) update(ctx context.Context) {
-	tm := time.NewTicker(g.timeOut)
+func (l *EpochListener) connect(ctx context.Context) error {
+	var (
+		opts rpcclient.WSOptions
+	)
 
+	for _, endpoint := range l.endpoints {
+		cl, err := rpcclient.NewWS(ctx, endpoint, opts)
+		if err != nil {
+			l.log.Info("create ws client", zap.Error(err), zap.String("endpoint", endpoint))
+			continue
+		}
+
+		if err = cl.Init(); err != nil {
+			l.log.Info("ws client init", zap.Error(err), zap.String("endpoint", endpoint))
+			continue
+		}
+
+		l.client = cl
+
+		l.log.Info("ws connected", zap.String("endpoint", endpoint))
+
+		return nil
+	}
+
+	return errors.New("all hosts failed to connect")
+}
+
+// ListenNotifications start listening notifications. Reconnects in case of disconnect.
+func (l *EpochListener) ListenNotifications(ctx context.Context) {
+	go func() {
+		bo := backoff.NewExponentialBackOff()
+
+		for {
+			select {
+			case <-ctx.Done():
+				l.log.Info("stopping. Context done")
+				return
+			default:
+			}
+
+			if err := l.connect(ctx); err != nil {
+				l.log.Error("connection failed", zap.Error(err))
+				time.Sleep(bo.NextBackOff())
```

Comment:
You need to reset it after successful connection. Otherwise it will grow even though there are no issues with establishing a new connection.